toolkit/offchain/src/collateral_selection.rs

```diff
@@ -0,0 +1,74 @@
+use ogmios_client::types::*;
+
+pub const MIN_REQUIRED_COLLATERAL: u64 = 5_000_000;
+
+// Largest-First selection algorithm based on https://cips.cardano.org/cip/CIP-0002#largest-first
+pub fn largest_first(
+	inputs_available: &[OgmiosUtxo],
+	amount_required: u64,
+	max_collateral_inputs: u32,
+	min_required_collateral: u64,
+) -> Result<Vec<&OgmiosUtxo>, &str> {
+	let amount_required = amount_required.max(min_required_collateral);
+	let amount_available: u64 = inputs_available.iter().map(|utxo| utxo.value.lovelace).sum();
+	if amount_available < amount_required {
+		return Err("InputValueInsufficientError");
+	}
+	let mut inputs_available_sorted = inputs_available.iter().collect::<Vec<_>>();
+	inputs_available_sorted
+		.sort_by(|utxo_a, utxo_b| utxo_a.value.lovelace.cmp(&utxo_b.value.lovelace));
+
+	let mut inputs_selected = Vec::new();
+	let mut sum = 0;
+	while sum < amount_required {
+		if let Some(utxo) = inputs_available_sorted.pop() {
+			sum += utxo.value.lovelace;
+			inputs_selected.push(utxo);
+		}
+	}
+
+	if inputs_selected.len() as u32 > max_collateral_inputs {
+		return Err("InputLimitExceededError");
```

Comment:
This won't be mapped, it will be displayed to the user.
How about: "Could not find min_required_collateral={min_required_collateral} lovelace in the max_collateral_inputs={max_collateral_inputs} inputs".